src/scores/probability/fss_impl.py

```diff
@@ -0,0 +1,528 @@
+"""
+This module contains methods related to the FSS score
+
+For an explanation of the FSS, and implementation considerations,
+see: [Fast calculation of the Fractions Skill Score][fss_ref]
+
+[fss_ref]:
+https://www.researchgate.net/publication/269222763_Fast_calculation_of_the_Fractions_Skill_Score
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import TYPE_CHECKING, Iterable, Optional, Tuple, Union
+
+import numpy as np
+import numpy.typing as npt
+import xarray as xr
+
+from scores import utils
+from scores.typing import XarrayLike
+
+# Note: soft keyword `type` only support from >=3.10
+# "tuple" of 64 bit floats
+f8x3 = np.dtype("f8, f8, f8")
+# Note: `TypeAlias` on variables for python <=3.9
+if TYPE_CHECKING:  # pragma: no cover
+    FssDecomposed = Union[np.ArrayLike, np.DtypeLike]
+else:  # pragma: no cover
+    FssDecomposed = npt.NDArray[f8x3]
+
+
+class DimensionError(ValueError):
+    """
+    Wrapper for specific coordinate/dimension mismatch errors
+    """
+
+    # better to be explicit because python doesn't have braces for scoping its
+    # control flow
+    pass  # pylint: disable=unnecessary-pass
+
+
+class FssComputeMethod(Enum):
+    """
+    Choice of compute backend for FSS.
+    - Standard
+        - NUMPY (default)
+    - Optimized
+        - NUMBA_NATIVE (unavailable)
+        - NUMBA_PARALLEL (unavailable)
+    - Experimental
+        - RUST_NATIVE (unavailable)
+        - RUST_OCL (unavailable)
+    """
+
+    NUMPY = 1  # (default)
+    NUMBA_NATIVE = 2
+    NUMBA_PARALLEL = 3
+    RUST_NATIVE = 4
+    RUST_OCL = 5
+
+
+def fss_2d(
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    *,  # Force keywords arguments to be keyword-only
+    threshold: np.float64,
+    window: Tuple[int, int],
+    spatial_dims: Tuple[str, str],
```

Comment:
Maybe just an iterable or collection?

Reply:
This is fine, and it's up to you. But if you want to support lists and sets as well as tuples, consider it. In fact, maybe you want a set because a tuple could have the same thing twice.

Reply:
Hmm, I chose tuple because, in this context it's:
- explicitly 2 fields
- immutable

Set does make sense for the uniqueness property, but I generally would use sets as read-only inputs if they are frozen or wrapped within an instantiated data type.

There seems to be a `frozenset` type I've not tried before. Maybe I could switch to that.

Reply:
I'm happy with it as is, if it's a deliberate choice for a reason.